### PR TITLE
[firebase-recaptcha] Add support for Firebase Phone Auth

### PR DESCRIPTION
**Pull request by IjzerenHein**
# Why

This PR makes it possible to use Firebase Phone Authentication through the use of a custom reCAPTCHA component. This because Firebase Phone Auth requires an Application Verifier object (reCAPTCHA) as an additional security measure to verify that the user is real and not a bot. This is an HTML widget and is otherwise not available on react-native.

[Full documentation preview](https://github.com/expo/expo/blob/%40hein/expo-firebase/recaptcha/docs/pages/versions/unversioned/sdk/firebase-recaptcha.md)

# How

This PR adds the `expo-firebase-recaptcha` package. This package provides a set of building blocks for creating a reCAPTCHA verifier and using that with your Firebase Phone authentication workflow.
It contains the following components/classes:

- `<FirebaseRecaptcha>` -- ReCAPTCHA widget that can be shown inline in your view
- `<FirebaseRecaptchaVerifierModal>` -- A modal that is shown when requesting reCAPTCHA verification. This is the easiest way to integrate reCAPTCHA verification with Firebase Phone Authentication.
- `FirebaseRecaptchaVerifier` -- A helper class for building custom reCAPTCHA verification workflows.

# Test Plan

The interactive reCAPTCHA workflow (modal) can be tested using the NCL. The modal itself uses the `FirebaseRecaptcha` component internally.

Additionally, unit-tests have been added to the package and `mounting` tests have been added to the test-suite.

## Recaptcha modal demo in NCL

ezgif com-video-to-gif (5)

## Full demo with Phone Auth

ezgif com-video-to-gif (4)

### CODE REVIEW

File: docs/pages/versions/unversioned/sdk/firebase-recaptcha.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+<PlatformsSection android emulator ios simulator />
+
+## Installation
+
+<InstallSection packageName="expo-firebase-recaptcha" />
+
+## Basic usage
+
+To get started[, read the offical Firebase phone-auth guide and **ignore all steps** that cover the reCAPTCHA configuration.](https://firebase.google.com/docs/auth/web/phone-auth)
```
Comment (brentvatne) [inline docs/pages/versions/unversioned/sdk/firebase-recaptcha.md]:
```suggestion
To get started, [read the offical Firebase phone-auth guide and **ignore all steps** that cover the reCAPTCHA configuration.](https://firebase.google.com/docs/auth/web/phone-auth)
```

File: docs/pages/versions/unversioned/sdk/firebase-recaptcha.md [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+  verificationId,
+  verificationCode
+);
+const authResult = await firebase.auth().signInWithCredential(credential);
+```
+
+
+## Example usage
+
+```tsx
```
Comment (brentvatne) [inline docs/pages/versions/unversioned/sdk/firebase-recaptcha.md]:
```suggestion
```js
```

need to investigate the root cause but syntax highlighting for jsx/tsx doesn't work properly in the docs in some cases so we're just using js for now

File: docs/pages/versions/unversioned/sdk/firebase-recaptcha.md [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    );
+  }
+}
+```
+
+## Customizing the appearance
+
+`<FirebaseRecaptchaVerifierModal>` has limited customisation options. You cannot change its appearance, but you can change the **title** and the **cancel-label**.
+
+```tsx
```
Comment (brentvatne) [inline docs/pages/versions/unversioned/sdk/firebase-recaptcha.md]:
```suggestion
```js
```

File: docs/pages/versions/unversioned/sdk/firebase-recaptcha.md [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+  ref={...}
+  firebaseConfig={...}
+  title='Prove you are human!'
+  cancelLabel='Close'
+/>
+```
+
+If you want a custom look & feel, then create your own `<Modal>` or display the `<FirebaseRecaptcha>` component inline in your screen.
+
+```tsx
```
Comment (brentvatne) [inline docs/pages/versions/unversioned/sdk/firebase-recaptcha.md]:
```suggestion
```js
```

File: docs/pages/versions/unversioned/sdk/firebase-recaptcha.md [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+- **firebaseVersion (string)** -- Optional version of the Firebase JavaScript SDK to load in the web-view. You can use this to load a custom or newer version. For example `version="6.8.0"`. 
+- **onVerify (function)** -- A callback that is invoked when reCAPTCHA has verified that the user is not a bot. The callback is provided with the reCAPTCHA token string. Example `onVerify={(recaptchaToken: string) => this.setState({recaptchaToken})}`.
+
+
+### `IFirebaseAuthApplicationVerifier`
+
+Interface describing a domain verification and abuse prevention verifier.
+
+```ts
+interface IFirebaseAuthApplicationVerifier {
```
Comment (brentvatne) [inline docs/pages/versions/unversioned/sdk/firebase-recaptcha.md]: @ide - any suggestions for the naming of this type? i can't recall any existing examples of using the `I` for `Interface` prefix in our other sdk packages

**Review by brentvatne — CHANGES_REQUESTED** [review]

File: docs/pages/versions/unversioned/sdk/firebase-recaptcha.md (hunk 5 above)
Comment (ide) [reply docs/pages/versions/unversioned/sdk/firebase-recaptcha.md]: Just name it `FirebaseAuthApplicationVerifier` and put whatever makes the concrete type concrete in the class's name. Or if there is nothing that distinguishes the concrete type, we don't need the interface.

Comment (IjzerenHein) [reply docs/pages/versions/unversioned/sdk/firebase-recaptcha.md]: I checked as well and couldn't find the `I` prefix or `Interface` suffix anywhere in the TS code of the packages. 

I've updated it to just `FirebaseAuthApplicationVerifier` 👍

**Comment by IjzerenHein** [comment]
One other thing I'd like to discuss. This library currently has a dependency on `expo-firebase-core`. It uses some TypeScript types from it and it's able to read the firebase web configuration from `app.json` for auto configuration purposes. This makes it dependent on SDK 37.

Would it make sense to be also to use this lib on SDK 36 or lower? I'd have to remove the dep and we'd loose the features mentioned above.

**Comment by IjzerenHein** [comment]
@brentvatne what do you reckon?
https://github.com/expo/expo/pull/7486#issuecomment-606057844

**Comment by brentvatne** [comment]
@IjzerenHein - i think it's fine for it to just be compatible with sdk37. imo it's a very easy update and people should do it :)

**Review by brentvatne — APPROVED** [review]

**Comment by brentvatne** [comment]
:shipit:

**Pull request merged**